hoomd/mpcd/methods.py

```diff
@@ -0,0 +1,107 @@
+    an MPCD solvent. The integration scheme is velocity Verlet with bounce-back
+    performed at the solid boundaries defined by a geometry, as in
+    `hoomd.mpcd.stream.BounceBack`. This gives a simple approximation of the
+    interactions required to keep a solute bounded in a geometry, and more complex
+    interactions can be specified, for example, by writing custom external fields.
+
+    Similar caveats apply to these methods as for `hoomd.mpcd.stream.BounceBack`.
+    In particular:
+
+    1. The simulation box is periodic, but the `geometry` may impose non-periodic
+       boundary conditions. You must ensure that the box is sufficiently large to
+       enclose the `geometry` and that all particles lie inside it, or an error will
+       be raised at runtime.
+    2. You must also ensure that particles do not self-interact through the periodic
+       boundaries. This is usually achieved for simple pair potentials by padding
+       the box size by the largest cutoff radius. Failure to do so may result in
+       unphysical interactions.
+    3. Bounce-back rules do not always enforce no-slip conditions at surfaces
+       properly. It may still be necessary to add additional "ghost" MD particles in
+       the surface to achieve the right boundary conditions and reduce density
+       fluctuations.
+
+    Warning:
+
+        This method does not support anisotropic integration because
+        torques are not computed for collisions with the boundary.
+        Rigid bodies will also not be treated correctly because the
+        integrator is not aware of the extent of the particles. The surface
+        reflections are treated as point particles. These conditions are too
+        complicated to validate easily, so it is the user's responsibility to
+        choose the `filter` correctly.
+
+    Attributes:
+        filter (hoomd.filter.filter_like): Subset of particles on which to apply
+            this method.
+
+        geometry (hoomd.mpcd.geometry.Geometry): Surface to bounce back from.
+
+    """
+
+    def __init__(self, filter, geometry):
+        super().__init__()
+
+        param_dict = ParameterDict(filter=ParticleFilter, geometry=Geometry)
+        param_dict.update(dict(filter=filter, geometry=geometry))
+        self._param_dict.update(param_dict)
+
+    def _attach_hook(self):
+        sim = self._simulation
+
+        self.geometry._attach(sim)
+
+        # try to find class in map, otherwise default to internal MPCD module
+        geom_type = type(self.geometry)
+        try:
+            class_info = self._class_map[geom_type]
+        except KeyError:
+            class_info = (_mpcd, "BounceBackNVE" + geom_type.__name__)
+        class_info = list(class_info)
+        if isinstance(sim.device, hoomd.device.GPU):
+            class_info[1] += "GPU"
+        class_ = getattr(*class_info, None)
+        assert class_ is not None, "Bounce back method for geometry not found"
```

Comment:
Shouldn't we raise the error every time this occurs, or is there a reason to limit it to running Python with asserts on?

Reply:
My thinking was that this error should only be raised for developers because they would need to implement and register the class. But, I can switch it to a regular exception if you think that's better.

Reply:
I am going to leave this as an assert because it is a developer error, and I couldn't pick a more meaningful built-in exception to raise instead.